simple_network_sim/sns_logger.py

```diff
@@ -0,0 +1,108 @@
+should - for consistency - be done before any information is written
+to the logger.
+
+To use the package-level logger, you might use
+
+from . import sns_logger
+logger = sns_logger.logger
+
+or
+
+from simple_network_sim import sns_logger
+logger = sns_logger.logger
+
+NOTE: By default, the logger writes to sys.stderr and only reports
+logger.WARNING and above.
+"""
+
+import logging
+import logging.config
+import sys
+
+from argparse import Namespace
+from pathlib import Path
+from typing import Optional
+
+def setup_logger(args: Optional[Namespace] = None) -> None:
+    """Create and return a logging.Logger instance.
+    
+    :param args: argparse.Namespace
+        args.logfile (pathlib.Path) is used to create a logfile if present
+        args.verbose and args.debug control logging level to sys.stderr
+    """
+    # Dictionary to define logging configuration
+    logconf = {
+        "version": 1,
+        "disable_existing_loggers": False,
+        "formatters": {
+            "standard": {
+                "format": "[%(asctime)s] [%(name)s] [%(levelname)s] %(message)s",
+                },
+            },
+        "handlers": {
+            "stderr": {
+                "level": "WARNING",
+                "class": "logging.StreamHandler",
+                "formatter": "standard",
+                "stream": "ext://sys.stderr"
+                },
+            },
+        "loggers": {
+            __package__: {
+                "handlers": ["stderr"],
+                "level": "DEBUG",
+                },
+            },
+        }
+
+    # If args.logpath is specified, add logfile
+    if args is not None and args.logfile is not None:
+        logdir = args.logfile.parents[0]
+        # If the logfile is going in another directory, we must
+        # create it/check if it's there
+        try:
+            if not logdir == Path.cwd():
+                logdir.mkdir(exist_ok=True)
+        except OSError:
+            logger.error("Could not create %s for logging", logdir, exc_info=True)
+            raise SystemExit(1)  # Substitute meaningful error code when known
+        # Add logfile configuration
+        logconf["handlers"]["logfile"] = {
+            "class": "logging.FileHandler",
+            "level": "INFO",
+            "formatter": "standard",
+            "filename": str(args.logfile),
+            "encoding": "utf8",
+            }
+        logconf["loggers"][__package__]["handlers"].append("logfile")
+
+    # Set STDERR/logfile levels if args.verbose/args.debug specified
+    if args is not None and args.verbose:
+        logconf["handlers"]["stderr"]["level"] = "INFO"
+    elif args is not None and args.debug:
+        logconf["handlers"]["stderr"]["level"] = "DEBUG"
+        logconf["handlers"]["logfile"]["level"] = "DEBUG"
+
+    # Configure logger
+    logging.config.dictConfig(logconf)
+
+# Provide a package-level logging.Logger object on import
+logger = logging.getLogger(__package__)
+setup_logger()
```

Comment:
I suggest just calling this in the main module explicitly. It's not really that important in our case (we don't intend for people to use any of the code as a library anyway), but it can be a bit surprising.

Reply:
Having this construction enables use of the common logger in each of the modules via an import. An alternative (and harder to maintain) mechanism might be to pass the `logger` explicitly as an argument to each and every function you want to use the logger for.

I understood the intent behind the logger to be to enable support for debugging and (implicitly) to report on issues/status when running via a CLI, as in `sampleUseOfModel.py`. Instantiating the logger only in `main()` doesn't provide support for debugging when modules are imported. This approach gives the same functionality as instantiating in `main()`, and supports `logger.debug()` messages - as opposed to `print()` statements (which I think came up in discussion).

Reply:
Hm, I think it's just a bit of pet peeve of mine, because I've dealt with libraries setting up the logger config in past and that and that was really annoying. But I think you're right. For this project, people are probably more interested in being able to import modules from the python shell and having it printing stuff rather than making the code into a library.